File: code/backend/java/adapter-rest/src/test/java/games/paths/adapters/rest/controller/match/MatchControllerTest.java
package games.paths.adapters.rest.controller.match;

import games.paths.adapters.rest.dto.MatchInfoResponse;
import games.paths.core.model.match.MatchCreateCommand;
import games.paths.core.model.match.MatchDetail;
import games.paths.core.model.match.MatchEventOption;
import games.paths.core.model.match.MatchLocationState;
import games.paths.core.model.match.MatchRegistryEntry;
import games.paths.core.model.match.MatchSummary;
import games.paths.core.port.match.MatchCommandPort;
import games.paths.core.port.match.MatchQueryPort;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.util.List;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

class MatchControllerTest {

    private MockMvc mockMvc;
    private MatchCommandPort commandPort;
    private MatchQueryPort queryPort;

    @BeforeEach
    void setUp() {
        commandPort = mock(MatchCommandPort.class);
        queryPort = mock(MatchQueryPort.class);
        mockMvc = MockMvcBuilders.standaloneSetup(new MatchController(commandPort, queryPort)).build();
    }

    private MockHttpServletRequestBuilder authed(MockHttpServletRequestBuilder b) {
        return b.requestAttr("userUuid", "user-uuid");
    }

    private MatchSummary summary() {
        MatchSummary s = new MatchSummary();
        s.setUuid("match-uuid");
        s.setStoryUuid("story-uuid");
        s.setDifficultyUuid("diff-uuid");
        s.setStatus("CREATED");
        s.setCurrentClock(0);
        s.setExpCost(5);
        s.setUserCreatorUuid("user-uuid");
        s.setName("name");
        s.setTsInsert("ts");
        s.setSinglePlayer(1);
        s.setCharacterTemplateUuid("char-tpl");
        s.setClassUuid("class-uuid");
        s.setTraitUuids(List.of("trait-1", "trait-2"));
        return s;
    }

    @Test
    void createMatch_unauthenticated_returns401() throws Exception {
        mockMvc.perform(post("/api/matches")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void createMatch_emptyBody_returns400() throws Exception {
        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{}"))
                .andExpect(status().isBadRequest());
    }

    @Test
    void createMatch_missingStory_returns400() throws Exception {
        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isBadRequest());
    }

    @Test
    void createMatch_missingDifficulty_returns400() throws Exception {
        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\"}"))
                .andExpect(status().isBadRequest());
    }

    @Test
    void createMatch_success_returns201() throws Exception {
        when(commandPort.createMatch(any())).thenReturn(summary());

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\",\"name\":\"n\"}"))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.uuid").value("match-uuid"))
                .andExpect(jsonPath("$.storyUuid").value("story-uuid"))
                .andExpect(jsonPath("$.singlePlayer").value(1))
                .andExpect(jsonPath("$.characterTemplateUuid").value("char-tpl"))
                .andExpect(jsonPath("$.classUuid").value("class-uuid"))
                .andExpect(jsonPath("$.traitUuids[0]").value("trait-1"));
    }

    @Test
    void createMatch_passesLoadoutToCommand() throws Exception {
        when(commandPort.createMatch(any())).thenReturn(summary());
        ArgumentCaptor<MatchCreateCommand> captor = ArgumentCaptor.forClass(MatchCreateCommand.class);

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\","
                                + "\"characterTemplateUuid\":\"ct\",\"classUuid\":\"cl\","
                                + "\"traitUuids\":[\"t1\",\"t2\"],\"singlePlayer\":0}"))
                .andExpect(status().isCreated());

        verify(commandPort).createMatch(captor.capture());
        MatchCreateCommand cmd = captor.getValue();
        org.junit.jupiter.api.Assertions.assertEquals("ct", cmd.getCharacterTemplateUuid());
        org.junit.jupiter.api.Assertions.assertEquals("cl", cmd.getClassUuid());
        org.junit.jupiter.api.Assertions.assertEquals(List.of("t1", "t2"), cmd.getTraitUuids());
        org.junit.jupiter.api.Assertions.assertEquals(0, cmd.getSinglePlayer());
    }

    @Test
    void createMatch_storyNotFound_returns404() throws Exception {
        when(commandPort.createMatch(any())).thenThrow(new MatchCommandPort.MatchCreationException(
                MatchCommandPort.MatchCreationException.Code.STORY_NOT_FOUND, "no story"));

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isNotFound())
                .andExpect(jsonPath("$.error").value("STORY_NOT_FOUND"));
    }

    @Test
    void createMatch_difficultyNotFound_returns404() throws Exception {
        when(commandPort.createMatch(any())).thenThrow(new MatchCommandPort.MatchCreationException(
                MatchCommandPort.MatchCreationException.Code.DIFFICULTY_NOT_FOUND, "no diff"));

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isNotFound());
    }

    @Test
    void createMatch_userNotFound_returns404() throws Exception {
        when(commandPort.createMatch(any())).thenThrow(new MatchCommandPort.MatchCreationException(
                MatchCommandPort.MatchCreationException.Code.USER_NOT_FOUND, "no user"));

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isNotFound());
    }

    @Test
    void createMatch_userBanned_returns403() throws Exception {
        when(commandPort.createMatch(any())).thenThrow(new MatchCommandPort.MatchCreationException(
                MatchCommandPort.MatchCreationException.Code.USER_BANNED, "banned"));

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isForbidden());
    }

    @Test
    void createMatch_maintenance_returns503() throws Exception {
        when(commandPort.createMatch(any())).thenThrow(new MatchCommandPort.MatchCreationException(
                MatchCommandPort.MatchCreationException.Code.MAINTENANCE_MODE, "maintenance"));

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isServiceUnavailable());
    }

    @Test
    void createMatch_storyHasNoLocations_returns400() throws Exception {
        when(commandPort.createMatch(any())).thenThrow(new MatchCommandPort.MatchCreationException(
                MatchCommandPort.MatchCreationException.Code.STORY_HAS_NO_LOCATIONS, "empty"));

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isBadRequest());
    }

    @Test
    void createMatch_invalidInput_returns400() throws Exception {
        when(commandPort.createMatch(any())).thenThrow(new MatchCommandPort.MatchCreationException(
                MatchCommandPort.MatchCreationException.Code.INVALID_INPUT, "bad"));

        mockMvc.perform(authed(post("/api/matches"))
                        .contentType(MediaType.APPLICATION_JSON)
                        .content("{\"storyUuid\":\"s\",\"difficultyUuid\":\"d\"}"))
                .andExpect(status().isBadRequest());
    }

    @Test
    void listMatches_unauthenticated_returns401() throws Exception {
        mockMvc.perform(get("/api/matches"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void listMatches_returnsArray() throws Exception {
        when(queryPort.listUserMatches("user-uuid")).thenReturn(List.of(summary()));
        mockMvc.perform(authed(get("/api/matches")))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[0].uuid").value("match-uuid"));
    }

    @Test
    void listAllMatches_returnsArray() throws Exception {
        when(queryPort.listAllMatches()).thenReturn(List.of(summary()));
        mockMvc.perform(get("/api/admin/matches"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$[0].uuid").value("match-uuid"));
    }

    @Test
    void listAllMatches_emptyArray() throws Exception {
        when(queryPort.listAllMatches()).thenReturn(List.of());
        mockMvc.perform(get("/api/admin/matches"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$").isArray());
    }

    @Test
    void getMatchInfo_unauthenticated_returns401() throws Exception {
        mockMvc.perform(get("/api/match/abc/info"))
                .andExpect(status().isUnauthorized());
    }

    @Test
    void getMatchInfo_notFound_returns404() throws Exception {
        when(queryPort.getMatchInfo("abc", "user-uuid")).thenReturn(null);
        mockMvc.perform(authed(get("/api/match/abc/info")))
                .andExpect(status().isNotFound());
    }

    @Test
    void getMatchInfo_returns200() throws Exception {
        MatchDetail detail = new MatchDetail();
        detail.setMatch(summary());
        detail.setCurrentLocationId(10L);
        detail.setCurrentLocationUuid("loc-uuid");
        detail.setCurrentLocationName("loc");
        MatchLocationState s = new MatchLocationState();
        s.setIdLocation(10L);
        s.setUuid("ls");
        s.setFlagAlreadyActived(0);
        s.setClockCounter(2);
        s.setName("ls-name");
        detail.setLocations(List.of(s));
        MatchRegistryEntry r = new MatchRegistryEntry();
        r.setUuid("r");
        r.setKey("k");
        r.setIntValue(1);
        detail.setRegistry(List.of(r));
        detail.setEvents(List.of(new MatchEventOption("ev", "n", "EVENT")));
        detail.setChoices(List.of(new MatchEventOption("ch", "n", "CHOICE")));
        when(queryPort.getMatchInfo("abc", "user-uuid")).thenReturn(detail);

        mockMvc.perform(authed(get("/api/match/abc/info")))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.match.uuid").value("match-uuid"))
                .andExpect(jsonPath("$.currentLocationId").value(10))
                .andExpect(jsonPath("$.locations[0].uuid").value("ls"))
                .andExpect(jsonPath("$.registry[0].key").value("k"))
                .andExpect(jsonPath("$.events[0].uuid").value("ev"))
                .andExpect(jsonPath("$.choices[0].uuid").value("ch"));
    }

    @Test
    void responseDtosFromModel_handleNullsAndConvertCorrectly() {
        // exercise the static factory methods on the response DTOs
        MatchInfoResponse nullCase = MatchInfoResponse.fromModel(null);
        org.junit.jupiter.api.Assertions.assertNull(nullCase);
    }
}

File: code/backend/php/tests/Unit/Adapter/Rest/Matches/MatchControllerTest.php
<?php

declare(strict_types=1);

namespace Games\Paths\Tests\Unit\Adapter\Rest\Matches;

use Games\Paths\Adapter\Rest\Matches\MatchController;
use Games\Paths\Core\Domain\Matches\MatchCreationException;
use Games\Paths\Core\Domain\Matches\MatchDetail;
use Games\Paths\Core\Domain\Matches\MatchEventOption;
use Games\Paths\Core\Domain\Matches\MatchLocationState;
use Games\Paths\Core\Domain\Matches\MatchRegistryEntry;
use Games\Paths\Core\Domain\Matches\MatchSummary;
use Games\Paths\Core\Port\Matches\MatchCommandPort;
use Games\Paths\Core\Port\Matches\MatchQueryPort;
use PHPUnit\Framework\TestCase;
use Slim\Psr7\Factory\ResponseFactory;
use Slim\Psr7\Factory\ServerRequestFactory;

class MatchControllerTest extends TestCase
{
    private $commandPort;
    private $queryPort;
    private MatchController $controller;
    private ResponseFactory $responseFactory;
    private ServerRequestFactory $requestFactory;

    protected function setUp(): void
    {
        $this->commandPort = $this->createMock(MatchCommandPort::class);
        $this->queryPort = $this->createMock(MatchQueryPort::class);
        $this->controller = new MatchController($this->commandPort, $this->queryPort);
        $this->responseFactory = new ResponseFactory();
        $this->requestFactory = new ServerRequestFactory();
    }

    private function summary(): MatchSummary
    {
        return new MatchSummary(
            'match-uuid', 'story-uuid', 'diff-uuid', 'n', 'CREATED', 0, 5, 'user-uuid', 'now',
            1, 'ct', 'cl', ['t1', 't2']
        );
    }

    private function detail(): MatchDetail
    {
        return new MatchDetail(
            match: $this->summary(),
            currentLocationId: 10,
            currentLocationUuid: 'loc',
            currentLocationName: 'loc-10',
            locations: [new MatchLocationState(10, 'ls', 0, 5, 'loc-10')],
            registry: [new MatchRegistryEntry('r', 'k', null, 1)],
            events: [new MatchEventOption('e', 'n', 'EVENT')],
            choices: [new MatchEventOption('c', 'n', 'CHOICE')]
        );
    }

    private function authedRequest(string $method, string $path, array $body = null)
    {
        $req = $this->requestFactory->createServerRequest($method, $path)
            ->withAttribute('userUuid', 'user-uuid');
        if ($body !== null) {
            $req = $req->withParsedBody($body);
        }
        return $req;
    }

    public function testCreateMatchUnauthenticated(): void
    {
        $request = $this->requestFactory->createServerRequest('POST', '/api/matches')
            ->withParsedBody(['storyUuid' => 's', 'difficultyUuid' => 'd']);
        $response = $this->responseFactory->createResponse();
        $result = $this->controller->createMatch($request, $response);
        $this->assertSame(401, $result->getStatusCode());
    }

    public function testCreateMatchEmptyBody(): void
    {
        $request = $this->authedRequest('POST', '/api/matches', []);
        $result = $this->controller->createMatch($request, $this->responseFactory->createResponse());
        $this->assertSame(400, $result->getStatusCode());
    }

    public function testCreateMatchMissingStory(): void
    {
        $request = $this->authedRequest('POST', '/api/matches', ['difficultyUuid' => 'd']);
        $result = $this->controller->createMatch($request, $this->responseFactory->createResponse());
        $this->assertSame(400, $result->getStatusCode());
    }

    public function testCreateMatchMissingDifficulty(): void
    {
        $request = $this->authedRequest('POST', '/api/matches', ['storyUuid' => 's']);
        $result = $this->controller->createMatch($request, $this->responseFactory->createResponse());
        $this->assertSame(400, $result->getStatusCode());
    }

    public function testCreateMatchSuccess(): void
    {
        $this->commandPort->method('createMatch')->willReturn($this->summary());
        $request = $this->authedRequest('POST', '/api/matches', [
            'storyUuid' => 's', 'difficultyUuid' => 'd', 'name' => 'n'
        ]);
        $result = $this->controller->createMatch($request, $this->responseFactory->createResponse());
        $this->assertSame(201, $result->getStatusCode());
        $body = json_decode((string)$result->getBody(), true);
        $this->assertSame('match-uuid', $body['uuid']);
        $this->assertSame(1, $body['singlePlayer']);
        $this->assertSame('cl', $body['classUuid']);
        $this->assertSame(['t1', 't2'], $body['traitUuids']);
    }

    public function testCreateMatchPassesLoadoutToCommand(): void
    {
        $captured = null;
        $this->commandPort->method('createMatch')->willReturnCallback(
            function ($command) use (&$captured) {
                $captured = $command;
                return $this->summary();
            }
        );
        $request = $this->authedRequest('POST', '/api/matches', [
            'storyUuid' => 's',
            'difficultyUuid' => 'd',
            'characterTemplateUuid' => 'ct',
            'classUuid' => 'cl',
            'traitUuids' => ['t1', 't2'],
            'singlePlayer' => 0,
        ]);
        $result = $this->controller->createMatch($request, $this->responseFactory->createResponse());
        $this->assertSame(201, $result->getStatusCode());
        $this->assertSame('ct', $captured->getCharacterTemplateUuid());
        $this->assertSame('cl', $captured->getClassUuid());
        $this->assertSame(['t1', 't2'], $captured->getTraitUuids());
        $this->assertSame(0, $captured->getSinglePlayer());
    }

    /**
     * @dataProvider errorCodes
     */
    public function testCreateMatchErrorCodes(string $code, int $expectedStatus): void
    {
        $this->commandPort->method('createMatch')
            ->willThrowException(new MatchCreationException($code, 'msg'));
        $request = $this->authedRequest('POST', '/api/matches', [
            'storyUuid' => 's', 'difficultyUuid' => 'd'
        ]);
        $result = $this->controller->createMatch($request, $this->responseFactory->createResponse());
        $this->assertSame($expectedStatus, $result->getStatusCode());
    }

    public static function errorCodes(): array
    {
        return [
            [MatchCreationException::STORY_NOT_FOUND, 404],
            [MatchCreationException::DIFFICULTY_NOT_FOUND, 404],
            [MatchCreationException::USER_NOT_FOUND, 404],
            [MatchCreationException::USER_BANNED, 403],
            [MatchCreationException::MAINTENANCE_MODE, 503],
            [MatchCreationException::STORY_HAS_NO_LOCATIONS, 400],
            [MatchCreationException::INVALID_INPUT, 400],
        ];
    }

    public function testCreateMatchUnknownErrorCodeFallsBackTo400(): void
    {
        $this->commandPort->method('createMatch')
            ->willThrowException(new MatchCreationException('UNKNOWN', 'msg'));
        $request = $this->authedRequest('POST', '/api/matches', [
            'storyUuid' => 's', 'difficultyUuid' => 'd'
        ]);
        $result = $this->controller->createMatch($request, $this->responseFactory->createResponse());
        $this->assertSame(400, $result->getStatusCode());
    }

    public function testListMatchesUnauthenticated(): void
    {
        $request = $this->requestFactory->createServerRequest('GET', '/api/matches');
        $result = $this->controller->listMatches($request, $this->responseFactory->createResponse());
        $this->assertSame(401, $result->getStatusCode());
    }

    public function testListMatchesReturnsArray(): void
    {
        $this->queryPort->method('listUserMatches')->willReturn([$this->summary()]);
        $request = $this->authedRequest('GET', '/api/matches');
        $result = $this->controller->listMatches($request, $this->responseFactory->createResponse());
        $this->assertSame(200, $result->getStatusCode());
        $body = json_decode((string)$result->getBody(), true);
        $this->assertCount(1, $body);
    }

    public function testListAllMatchesReturnsArray(): void
    {
        $this->queryPort->method('listAllMatches')->willReturn([$this->summary()]);
        $request = $this->requestFactory->createServerRequest('GET', '/api/admin/matches');
        $result = $this->controller->listAllMatches($request, $this->responseFactory->createResponse());
        $this->assertSame(200, $result->getStatusCode());
        $body = json_decode((string)$result->getBody(), true);
        $this->assertCount(1, $body);
        $this->assertSame('match-uuid', $body[0]['uuid']);
    }

    public function testListAllMatchesEmpty(): void
    {
        $this->queryPort->method('listAllMatches')->willReturn([]);
        $request = $this->requestFactory->createServerRequest('GET', '/api/admin/matches');
        $result = $this->controller->listAllMatches($request, $this->responseFactory->createResponse());
        $this->assertSame(200, $result->getStatusCode());
        $this->assertSame([], json_decode((string)$result->getBody(), true));
    }

    public function testGetMatchInfoUnauthenticated(): void
    {
        $request = $this->requestFactory->createServerRequest('GET', '/api/match/abc/info');
        $result = $this->controller->getMatchInfo($request, $this->responseFactory->createResponse(), ['uuidMatch' => 'abc']);
        $this->assertSame(401, $result->getStatusCode());
    }

    public function testGetMatchInfoMissingUuid(): void
    {
        $request = $this->authedRequest('GET', '/api/match//info');
        $result = $this->controller->getMatchInfo($request, $this->responseFactory->createResponse(), []);
        $this->assertSame(400, $result->getStatusCode());
    }

    public function testGetMatchInfoNotFound(): void
    {
        $this->queryPort->method('getMatchInfo')->willReturn(null);
        $request = $this->authedRequest('GET', '/api/match/abc/info');
        $result = $this->controller->getMatchInfo($request, $this->responseFactory->createResponse(), ['uuidMatch' => 'abc']);
        $this->assertSame(404, $result->getStatusCode());
    }

    public function testGetMatchInfoSuccess(): void
    {
        $this->queryPort->method('getMatchInfo')->willReturn($this->detail());
        $request = $this->authedRequest('GET', '/api/match/abc/info');
        $result = $this->controller->getMatchInfo($request, $this->responseFactory->createResponse(), ['uuidMatch' => 'abc']);
        $this->assertSame(200, $result->getStatusCode());
        $body = json_decode((string)$result->getBody(), true);
        $this->assertSame('match-uuid', $body['match']['uuid']);
        $this->assertSame(10, $body['currentLocationId']);
        $this->assertCount(1, $body['locations']);
        $this->assertCount(1, $body['registry']);
    }
}
